src/Data/AudienceTree.elm

```diff
@@ -0,0 +1,361 @@
+module Data.AudienceTree exposing
+    ( AudienceTree
+    , DuplicatedIdFolder
+    , Errors
+    , audiences
+    , children
+    , fromAudiencesAndFolders
+    , name
+    )
+
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Dict exposing (Dict)
+
+
+type AudienceTree
+    = AudienceTree AudienceTreeInternal
+
+
+type alias AudienceTreeInternal =
+    { name : String
+    , children : List AudienceTree
+    , audiences : List String
+    }
+
+
+children : AudienceTree -> List AudienceTree
+children (AudienceTree a) =
+    a.children
+
+
+name : AudienceTree -> String
+name (AudienceTree a) =
+    a.name
+
+
+audiences : AudienceTree -> List String
+audiences (AudienceTree a) =
+    a.audiences
+
+
+type alias Errors =
+    { duplicatedFolderIds : List DuplicatedIdFolder
+    , missingFolderId : List MissingParentAudience
+    , missingFolderParentId : List MissingParentFolder
+    }
+
+
+{-| Build the audience tree from the lists of audiences and audience folders.
+It will always succeed to build some Tree, even if some ids are inconsistent.
+E.g. if an audience folder references a non-existing folder, this audience is directly
+added to the root, and this audience is added to the `missingFolderId` field.
+-}
+fromAudiencesAndFolders : List Audience -> List AudienceFolder -> ( AudienceTree, Errors )
+fromAudiencesAndFolders audiences_ folders =
+    let
+        step3result =
+            step1 folders
+                |> step2 audiences_
+                |> step3
+    in
+    ( step3result.audienceTree
+    , { duplicatedFolderIds = step3result.duplicatedFolderIds
+      , missingFolderId = step3result.missingFolderId
+      , missingFolderParentId = step3result.missingFolderParentId
+      }
+    )
+
+
+
+-- STEP 1
+--
+-- Transform the list of `AudienceFolder` into a dictionnary `Id -> PendingFolder`
+-- each `PendingFolder` being "empty": only name and parent are provided.
+--
+-- It would be problematic if multiple folders share the same id `I`. If an audience
+-- belongs to `I`, to what folder should we put it? If we encounter
+-- such a case, only the first folder is kept and we keep track of those problematic folders
+-- in the `duplicatedFolderIds` list.
+
+
+type alias Step1Result =
+    { root : PendingRootFolder
+    , otherFolders : Dict Int PendingFolder
+    , duplicatedFolderIds : List DuplicatedIdFolder
```

Comment:
This is super overkill I would say. It's nice, but I thinks (as I mentioned already) that this has to be handled on BE. Especially if you have control over it (API is own by your team or team in your company). 
And also it generates 4 extra lines of comments and I don't like too much comments :)